akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameLogger.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2009-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.ws
+
+import akka.NotUsed
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.stream.Attributes
+import akka.stream.Attributes.LogLevels
+import akka.stream.scaladsl.{ BidiFlow, Flow }
+import akka.util.ByteString
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[ws] object FrameLogger {
+  final val maxBytes = 16
+
+  def logFramesIfEnabled(shouldLog: Boolean): BidiFlow[FrameEvent, FrameEvent, FrameEvent, FrameEvent, NotUsed] =
+    if (shouldLog) bidi
+    else BidiFlow.identity
+
+  def bidi: BidiFlow[FrameEvent, FrameEvent, FrameEvent, FrameEvent, NotUsed] =
+    BidiFlow.fromFlows(
+      Flow[FrameEvent].log(s"${Console.RED}DOWN${Console.RESET}", FrameLogger.logEvent),
+      Flow[FrameEvent].log(s"${Console.GREEN} UP ${Console.RESET}", FrameLogger.logEvent))
+      .addAttributes(Attributes(LogLevels(Logging.DebugLevel, Logging.DebugLevel, Logging.DebugLevel)))
+
+  def logEvent(frameEvent: FrameEvent): String = {
+    case class LogEntry(frameType: String, data: String)
+
+    def hex(bytes: ByteString): String = {
+      val num = math.min(maxBytes, bytes.size)
+      val ellipsis = if (num < bytes.size) s" [... ${bytes.size - num} more bytes]" else ""
+      bytes
+        .take(num)
+        .map(_ formatted "%02x")
+        .mkString(" ") + ellipsis
+    }
+
+    def entryForFrame(frameEvent: FrameEvent): LogEntry =
+      frameEvent match {
+        case FrameStart(_, data) => LogEntry("FRAMESTART", hex(data))
```

Comment:
The important thing here is to also log the frame header:

```
private[http] final case class FrameHeader(
  opcode: Protocol.Opcode,
  mask:   Option[Int],
  length: Long,
  fin:    Boolean,
  rsv1:   Boolean         = false,
  rsv2:   Boolean         = false,
  rsv3:   Boolean         = false)
```

 * The Opcode should be a textual representation of the opcode, maybe just the `toString.toUperString` for now. It can be used instead of `FRAMESTART`.
 * At the point where it's logged, no mask should be set, so that can be ignored.
 * Length should be shown before the actual data
 * The flags can be simple strings that are only shown if the flag is set, so add `FIN` `RSV1`, etc when the flag is set and leave it out otherwise.

Reply:
Also it should add some indication if all data is contained in that event or not (i.e. the value of `FrameStart.lastPart`)

Reply:
Updated the PR, The logs now look like below: 
<img width="1792" alt="Screenshot 2021-02-03 at 12 09 16 AM" src="https://user-images.githubusercontent.com/54163056/106646868-2f7a6080-65b4-11eb-971e-fb1e5a4b8d40.png">

Reply:
The `lastPart` / `T` / `F` rendering was redundant as it should already included in the flags rendering.